src/Temporalio/Worker/WorkflowCodecHelper.cs

```diff
@@ -0,0 +1,279 @@
+using System.Linq;
+using System.Threading.Tasks;
+using Google.Protobuf.Collections;
+using Temporalio.Api.Common.V1;
+using Temporalio.Bridge.Api.ActivityResult;
+using Temporalio.Bridge.Api.ChildWorkflow;
+using Temporalio.Bridge.Api.WorkflowActivation;
+using Temporalio.Bridge.Api.WorkflowCommands;
+using Temporalio.Bridge.Api.WorkflowCompletion;
+using Temporalio.Converters;
+
+namespace Temporalio.Worker
+{
+    /// <summary>
+    /// Utilities to encode/decode workflow protobufs.
+    /// </summary>
+    internal static class WorkflowCodecHelper
+    {
+        /// <summary>
+        /// Encode the completion.
+        /// </summary>
+        /// <param name="codec">Codec to use.</param>
+        /// <param name="comp">Completion to encode.</param>
+        /// <returns>Task for completion.</returns>
+        internal static async Task EncodeAsync(
+            IPayloadCodec codec, WorkflowActivationCompletion comp)
+        {
+            switch (comp.StatusCase)
+            {
+                case WorkflowActivationCompletion.StatusOneofCase.Failed:
+                    if (comp.Failed.Failure_ != null)
+                    {
+                        await codec.EncodeFailureAsync(comp.Failed.Failure_).ConfigureAwait(false);
+                    }
+                    break;
+                case WorkflowActivationCompletion.StatusOneofCase.Successful:
+                    foreach (var cmd in comp.Successful.Commands)
+                    {
+                        await EncodeAsync(codec, cmd).ConfigureAwait(false);
+                    }
+                    break;
+            }
+        }
+
+        /// <summary>
+        /// Decode the activation.
+        /// </summary>
+        /// <param name="codec">Codec to use.</param>
+        /// <param name="act">Activation to decode.</param>
+        /// <returns>Task for completion.</returns>
+        internal static async Task DecodeAsync(IPayloadCodec codec, WorkflowActivation act)
+        {
+            foreach (var job in act.Jobs)
+            {
+                switch (job.VariantCase)
```

Comment:
Hrmm, I think I missed headers in Python (and the aforementioned reflection test would catch these for .NET). Will fix here, but what should I do about Python? Leave headers unencoded for compat reasons?

Reply:
Headers should be encoded, they may contain sensitive information.

Reply:
I am asking about what to do about Python knowing it already exists unencoded today.

Reply:
I meant for Python

Reply:
> Headers should be encoded, they may contain sensitive information.
> I meant for Python

We know that headers should be encoded. The question is what to do for Python when users are already using unencoded headers today. The question is about compatibility not about whether they should be encoded. I can take this out of this PR and we can discuss what we should do in Python since the question is not coming across here.